### PR TITLE
Allow dynamic key reloading when having inactive keys (imported & derived)

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

Supersedes #8076

- Updating keys in the keystore will trigger a restart of `WaitForActivation`so that the new key can be fetched. Currently no new keys are fetched if all existing keys are inactive.
- Reload doesn't work when the wallet is empty. This PR fixes the issue.

**Which issues(s) does this PR fix?**

Part of #7887

**Other notes for review**

This fix applies only to imported and derived wallets. Remote wallets require much more work, such as an additional RPC endpoint in the signer and the ability to reach that endpoint, and are out of scope of this PR.

### CODE REVIEW

File: validator/client/runner.go [hunk 1]
```diff
@@ -67,9 +71,17 @@ func run(ctx context.Context, v Validator) {
...
 		log.Fatalf("Could not determine if beacon node synced: %v", err)
 	}
-	if err := v.WaitForActivation(ctx); err != nil {
+
+	val, ok := v.(*validator)
+	if !ok {
+		log.Fatalf("Could not wait for validator activation: could not instantiate validator")
+	}
+	accountsChangedChan := make(chan struct{})
+	go val.handleAccountsChanged(ctx, accountsChangedChan)
```
Comment (rauljordan) [inline validator/client/runner.go:80]: Don't feel too great about this, because we wanted to keep this `run` function abstract, using the interface only. Doing this type assertion goes against the design of this function

File: validator/client/runner.go [hunk 2]
```diff
@@ -172,3 +184,35 @@ func handleAssignmentError(err error, slot uint64) {
 		log.WithField("error", err).Error("Failed to update assignments")
 	}
 }
+
+func (v *validator) handleAccountsChanged(ctx context.Context, accountsChangedChan chan struct{}) {
+	validatingPubKeysChan := make(chan [][48]byte, 1)
```
Comment (rauljordan) [inline validator/client/runner.go:189]: We should defer close this channel

File: validator/client/runner.go [hunk 3]
```diff
@@ -172,3 +184,35 @@ func handleAssignmentError(err error, slot uint64) {
...
+	var sub event.Subscription
+	importedKeymanager, ok := v.keyManager.(*imported.Keymanager)
```
Comment (rauljordan) [inline validator/client/runner.go:191]: Use a switch here and throw an error or return if the keymanager is remote perhaps

Comment (rauljordan) [reply validator/client/runner.go:191]: You can do a type switch
```go
switch v.keyManager.(type) {
}
```

File: validator/client/wait_for_activation.go [hunk 4]
```diff
@@ -66,44 +66,52 @@ func (v *validator) WaitForActivation(ctx context.Context) error {
...
+				break
+			}
+			// If context is canceled we return from the function.
+			if ctx.Err() == context.Canceled {
+				return errors.Wrap(ctx.Err(), "context has been canceled so shutting down the loop")
+			}
+			if err != nil {
+				traceutil.AnnotateError(span, err)
+				attempts := getStreamAttempts(ctx)
+				log.WithError(err).WithField("attempts", attempts).
```
Comment (rauljordan) [inline validator/client/wait_for_activation.go:89]: Debug log perhaps?

**Review by rauljordan — COMMENTED** [review]
Approach is good save for the placement of some code, left a few comments

File: validator/client/runner.go (hunk 2 above)
Comment (rkapka) [reply validator/client/runner.go:189]: Thanks, I will also close the channel in `service.recheckValidatingKeysBucket`.

File: validator/client/wait_for_activation.go (hunk 4 above)
Comment (rkapka) [reply validator/client/wait_for_activation.go:89]: I did not touch this code. I can change it to debug, but I guess the reason for `WithError` was that we have `err != nil`.

File: validator/client/runner.go (hunk 3 above)
Comment (rkapka) [reply validator/client/runner.go:191]: Thanks. totally forgot about the switch option :-)

As for throwing an error, I cannot do that. The runner will execute this function with all types of wallets, and therefore throwing an error for remote wallets would make them unusable.

File: validator/client/runner.go (hunk 1 above)
Comment (rkapka) [reply validator/client/runner.go:80]: Yeah, this doesn't look too good. I will add a new function to the interface, so that no type assertions are required.

File: validator/client/runner.go [hunk 5]
```diff
@@ -67,9 +73,13 @@ func run(ctx context.Context, v Validator) {
 	if err := v.WaitForSync(ctx); err != nil {
 		log.Fatalf("Could not determine if beacon node synced: %v", err)
 	}
-	if err := v.WaitForActivation(ctx); err != nil {
+
+	accountsChangedChan := make(chan struct{})
```
Comment (rauljordan) [inline validator/client/runner.go:77]: Any reason this is an unbuffered channel? It will block on write if there is no listener ready (although seems like it's ok in this case)

File: validator/client/runner.go [hunk 6]
```diff
@@ -172,3 +182,33 @@ func handleAssignmentError(err error, slot uint64) {
 		log.WithField("error", err).Error("Failed to update assignments")
 	}
 }
+
+func handleAccountsChanged(ctx context.Context, v Validator, accountsChangedChan chan struct{}) {
```
Comment (prestonvanloon) [inline validator/client/runner.go:186]: Any test for this function?

**Review by prestonvanloon — COMMENTED** [review]
Overall, looks good to me, juts had a question about new logic in runner.go missing a test.

File: validator/client/runner.go (hunk 6 above)
Comment (rkapka) [reply validator/client/runner.go:186]: Good idea. This code is executed when testing `WaitForActivation`, but it is beneficial to test this function in isolation.

Comment (rkapka) [reply validator/client/runner.go:186]: After some consideration I am in favor of postponing these tests until dynamic key reloading for the remote keymanager is ready. This will allow to define `SubscribeAccountChanges` as an interface function and make testing much easier.

File: validator/accounts/wallet_create.go [hunk 7]
```diff
@@ -35,10 +37,6 @@ func CreateAndSaveWalletCli(cliCtx *cli.Context) (*wallet.Wallet, error) {
 	if err != nil {
 		return nil, err
 	}
-	if keymanagerKind == keymanager.Imported {
-		log.Warning("Creating an imported wallet does not provide anything beneficial. " +
-			"This option will be removed in v1.0.0. Please use `validator accounts import` instead.")
-	}
```
Comment (prestonvanloon) [inline validator/accounts/wallet_create.go:41]: How is this relevant to your dynamic key reloading changes? 
Should it be removed? cc: @rauljordan

Comment (prestonvanloon) [reply validator/accounts/wallet_create.go:41]: Why are there changes in this file at all? I thought this PR was to fix operations with dynamic key reloading on existing wallets.

File: validator/client/runner.go (hunk 6 above)
Comment (prestonvanloon) [reply validator/client/runner.go:189]: I'm not in favor of postponing tests on such mission critical functionality. 

If something is blocking tests for this PR, wait until that issue has been resolved then add tests before proposing these changes.

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
I'm concerned about the scope of this change and the missing tests on the critical functionality of the validator's runner.go.

File: validator/accounts/wallet_create.go (hunk 7 above)
Comment (rkapka) [reply validator/accounts/wallet_create.go:41]: This was removed because we now allow empty wallets. Previously it did not make sense to have one. It has been some time since this change, so I don't remember the exact details. If I'm not mistaken, it was a request from Bison Trails.

@rauljordan Can you confirm?

**Comment by rkapka** [comment]
`TestWaitForActivation` tests run through `handleAccountsChanged`, but there are no unit tests of this function in a vacuum. I will try to write some, it's just that it will probably require a complicated wallet setup. I cannot use a mock here because the function does type assertions.

File: validator/accounts/wallet_create.go (hunk 7 above)
Comment (rauljordan) [reply validator/accounts/wallet_create.go:41]: Staking as a service providers prefer to create an empty wallet and then dynamically import keys as needed in their cloud deployments. We should probably move that to a separate PR IMO to keep the diff small, however.

File: validator/accounts/accounts.go [hunk 8]
```diff
@@ -8,7 +8,11 @@ import (
 
 var log = logrus.WithField("prefix", "accounts")
 
-var msgKeymanagerNotSupported = "keymanager kind not supported: %s"
+var (
+	msgKeymanagerNotSupported = "keymanager kind not supported: %s"
+	// MsgCouldNotInitializeKeymanager informs about failed keymanager initialization
+	MsgCouldNotInitializeKeymanager = "could not initialize keymanager"
```
Comment (rauljordan) [inline validator/accounts/accounts.go:14]: Should be ErrCouldNotInitializeKeymanager

File: validator/client/runner.go [hunk 9]
```diff
@@ -175,3 +185,34 @@ func handleAssignmentError(err error, slot uint64) {
...
+	validatingPubKeysChan := make(chan [][48]byte, 1)
+	var sub event.Subscription
+	switch km := v.GetKeymanager().(type) {
```
Comment (rauljordan) [inline validator/client/runner.go:192]: Should this function instead take in a keymanager iface as an argument instead of the `validator` ? Also, we can consider adding SubscribeAccountChanges as a method for the interface, and for remote wallets, simply return error cannot subscribe to account changes. This will make it easier for you to avoid having to do this type assertion. Also, it will make this function really easy to test as all you will need is a mock for the small keymanager interface and can be unit tested. I don't see why tests are blocked for this function if we change these items

File: validator/client/runner.go [hunk 10]
```diff
@@ -175,3 +185,34 @@ func handleAssignmentError(err error, slot uint64) {
...
+		sub.Unsubscribe()
+		close(validatingPubKeysChan)
+	}()
+
+	for {
+		select {
+		case <-validatingPubKeysChan:
+			accountsChangedChan <- struct{}{}
+		case <-ctx.Done():
+			return
```
Comment (rauljordan) [inline validator/client/runner.go:211]: There's also a sub.Err() case I believe we should check

File: validator/accounts/wallet_create.go (hunk 7 above)
Comment (rkapka) [reply validator/accounts/wallet_create.go:41]: Agreed that we can move it to a separate PR and merge it before this one.

File: validator/client/runner.go (hunk 9 above)
Comment (rkapka) [reply validator/client/runner.go:192]: We can't throw an error because it will make remote wallets unusable - you would get an error every time. We could make it a no-op, though. That would indeed simplify everything.

File: validator/accounts/accounts.go (hunk 8 above)
Comment (rkapka) [reply validator/accounts/accounts.go:14]: Fixed in #8251

**Comment by rkapka** [comment]
Blocked by #8251

File: validator/accounts/accounts_list_test.go [hunk 11]
```diff
@@ -9,6 +9,8 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/prysmaticlabs/prysm/shared/event"
+
```
Comment (rauljordan) [inline validator/accounts/accounts_list_test.go:13]:
```suggestion
```

**Review by rauljordan — APPROVED** [review]

**Review by rauljordan — APPROVED** [review]

**Review by prestonvanloon — APPROVED** [review]
